### PR TITLE
Generator: reduce amount of intermediate allocations

### PR DESCRIPTION
**Pull request by Kijewski**
No need to allocate a String, just to write it into another String afterwards. This PR replaces the usages for `format!()` with `format_args!()`.

### CODE REVIEW

File: rinja_derive/src/generator.rs [hunk 1]
```diff
@@ -122,17 +123,17 @@ impl<'a> Generator<'a> {
...
-        buf.writeln(";");
-
-        buf.writeln("const SIZE_HINT: ::std::primitive::usize = ");
-        buf.writeln(&format!("{size_hint}"));
-        buf.writeln(";");
-
-        buf.writeln("const MIME_TYPE: &'static ::std::primitive::str = ");
-        buf.writeln(&format!("{:?}", &self.input.mime_type));
-        buf.writeln(";");
+        buf.writeln(format_args!(
```
Comment (GuillaumeGomez) [inline rinja_derive/src/generator.rs:126]: Couldn't it be replaced with the [`writeln` macro](https://doc.rust-lang.org/nightly/std/macro.writeln.html)?

Comment (Kijewski) [reply rinja_derive/src/generator.rs:126]: [`format_args_nl!()`](https://doc.rust-lang.org/nightly/std/macro.format_args_nl.html) is not stable, yet.

**Comment by GuillaumeGomez** [comment]
That makes me think that we should add a way to bench performance impact somehow (not memory, too unlikely to be useful).

File: rinja_derive/src/generator.rs (hunk 1 above)
Comment (Kijewski) [reply rinja_derive/src/generator.rs:126]: Ah, I think I misunderstood you. Answering in the next comment.

File: rinja_derive/src/generator.rs [hunk 2]
```diff
@@ -1169,10 +1169,7 @@ impl<'a> Generator<'a> {
                 let id = self.named;
                 self.named += 1;
 
-                buf_expr.write(&format!("expr{id} = "));
-                buf_expr.write("&");
-                buf_expr.write(entry.key());
-                buf_expr.writeln(",");
+                buf_expr.write(format_args!("expr{id} = &{},", entry.key()));
```
[the remark this answers is not in the document]
Comment (Kijewski) [reply rinja_derive/src/generator.rs:1172]: I am not sure, but I think that would have a negative performance impact. We only work on Strings, but `write!()` works on `&dyn Write`. I don't know if the compiler is still able to optimize the code, not to employ vtables, is it?

**Comment by Kijewski** [comment]
Is it possible to (repeatedly) call [`rinja_generator::derive_template()`](https://github.com/rinja-rs/rinja/blob/master/rinja_derive/src/lib.rs#L25) in "normal" code? If so, then I guess benchmarking would be easy. Otherwise, I guess it would be mostly a benchmark for your harddisk and the filesystem cache. :/

**Comment by GuillaumeGomez** [comment]
I thought we could use something like we have [here](https://github.com/rinja-rs/rinja/blob/master/rinja_derive/src/tests.rs#L7). We can do an `include_str!` and then call the parser on it in the bench I think?

File: rinja_derive/src/generator.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply rinja_derive/src/generator.rs:1172]: Well, godbolt could answer that, let me check.

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply rinja_derive/src/generator.rs:1172]: I tried with this code:

```rust
use std::fmt::Write;

// If you use `main()`, declare it as `pub` to see it in the output:
pub fn main() -> usize {
    let mut s = String::new();
    let a = "a";
    let b = 12;

    s.write_fmt(format_args!("{a} {b}"));
    write!(s, "{a} {b}");
    0
}
```

I compiled with `-C opt-level=3`. So for `s.write_fmt(format_args!("{a} {b}"));` we have:

```asm
        mov     qword ptr [rsp + 40], rbx
        lea     r14, [rip + <&T as core::fmt::Display>::fmt::h0e293c54e2c02aa4]
        mov     qword ptr [rsp + 48], r14
        lea     r15, [rsp + 12]
        mov     qword ptr [rsp + 56], r15
        mov     r12, qword ptr [rip + core::fmt::num::imp::<impl core::fmt::Display for i32>::fmt::h4d2c4c91afac718c@GOTPCREL]
        mov     qword ptr [rsp + 64], r12
```

and for `write!(s, "{a} {b}");` we have:

```asm
        call    qword ptr [rip + core::fmt::write::hc090a2ffd6b28c4a@GOTPCREL]
        mov     qword ptr [rsp + 40], rbx
        mov     qword ptr [rsp + 48], r14
        mov     qword ptr [rsp + 56], r15
        mov     qword ptr [rsp + 64], r12
```

So at this point, I think only a benchmark would be able to differentiate both.

**Review by GuillaumeGomez — APPROVED** [review]
I think it'd be interesting to give a try to `write!` and `writeln!` macros but the current code is already better than the existing so can be done in a follow-up!

**Comment by GuillaumeGomez** [comment]
I opened #12 and #13 so it's not forgotten. I think it's really minor improvements so not a priority at all, but could still be interesting. :)

**Pull request merged**